linkerd/app/core/src/retry.rs

```diff
@@ -183,9 +134,20 @@ where
         &self,
         req: http::Request<A>,
     ) -> Either<Self::RetryRequest, http::Request<A>> {
-        if self.can_retry(&req) {
-            return Either::A(req.map(|body| ReplayBody::new(body, MAX_BUFFERED_BYTES)));
-        }
-        Either::B(req)
+        let (head, body) = req.into_parts();
+        let replay_body = match ReplayBody::try_new(body, MAX_BUFFERED_BYTES) {
+            Ok(body) => body,
+            Err(body) => {
+                tracing::debug!(
+                    size = body.size_hint().lower(),
+                    "Body is too large to buffer"
+                );
+                return Either::B(http::Request::from_parts(head, body));
+            }
+        };
+
+        // The body may still be too large to be buffered if the body's length was not known.
+        // `ReplayBody` handles this gracefully.
+        Either::A(http::Request::from_parts(head, replay_body))
```

Comment:
nit, take it or leave it: the early return here feels slightly convoluted, can't this just be
```suggestion
        match ReplayBody::try_new(body, MAX_BUFFERED_BYTES) {
            Ok(replay_body) => {
                // The body may still be too large to be buffered if the body's length was not known.
                // `ReplayBody` handles this gracefully.
                Either::A(http::Request::from_parts(head, replay_body))
            >,
            Err(body) => {
                tracing::debug!(
                    size = body.size_hint().lower(),
                    "Body is too large to buffer"
                );
                Either::B(http::Request::from_parts(head, body))
            }
        }
```

Reply:
yeah, I had it that way initially. but it feels like the secondary case is exceptional, so it feels appropriate to early return. don't feel strongly either way, though.